TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilderNodeDispatcher.cs

```diff
@@ -0,0 +1,1322 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using TypeCobol.Compiler.CodeElements;
+using TypeCobol.Compiler.Nodes;
+using TypeCobol.Compiler.Parser;
+
+namespace TypeCobol.Compiler.CupParser.NodeBuilder
+{
+    /// <summary>
+    /// A Node Dispatcher on a ProgramClassBuilder that is able to also dispatch building actions.
+    /// </summary>
+    public class ProgramClassBuilderNodeDispatcher : NodeDispatcher, IProgramClassBuilder
+    {
+        private IList<ProgramClassBuilderNodeListener> _builderListeners = null;
+
+        public ProgramClassBuilderNodeDispatcher()
+        {
+        }
+
+        /// <summary>
+        /// Overriden AddListener to filter those that are ProgramClassBuilderNodeListener instance.
+        /// So that we can build a specific list, and avoid dynamic type checking at dispatch time.
+        /// </summary>
+        /// <param name="listener">The listener to be added</param>
+        protected override void AddListener(NodeListener listener)
+        {
+            base.AddListener(listener);
+            if (listener is ProgramClassBuilderNodeListener nodeListener)
+            {
+                if (_builderListeners == null)
+                {
+                    _builderListeners = new List<ProgramClassBuilderNodeListener>();
+                }
+                _builderListeners.Add(nodeListener);
+            }
+        }
+
+        public virtual void OnLevel1Definition(DataDefinition level1Node)
+        {
+            if (_builderListeners != null) foreach (var l in _builderListeners)
+                {
+                    l.OnLevel1Definition(level1Node);
+                }
+        }
+
+        public virtual void Enter(Node node)
+        {
+            if (_builderListeners != null) foreach (var l in _builderListeners)
+                {
+                    l.Enter(node);
+                }
+        }
+
+        public virtual void Exit(Node node)
+        {
+            if (_builderListeners != null) foreach (var l in _builderListeners)
+                {
+                    l.Exit(node);
+                }
+        }
+
+        public void AddNextSentenceStatement([NotNull] NextSentenceStatement stmt)
+        {
+            if (_builderListeners != null) foreach (var l in _builderListeners)
+                {
+                    l.AddNextSentenceStatement(stmt);
+                }
+        }
+
+        public void CheckStartSentenceLastStatement()
+        {
+            if (_builderListeners != null) foreach (var l in _builderListeners)
+                {
+                    l.CheckStartSentenceLastStatement();
+                }
+        }
+
+        public void EndAddStatementConditional(AddStatementEnd end = null)
```

Comment:
It can, because of the cup specification grammar:
addStatementConditional ::=
		addStatement
		{: my_parser.Builder.EndAddStatementConditional(); :}